File: src/option_pricing.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include "../include/option_pricing.h"

/**
 * @brief Thread-safe, reentrant European Option pricer (CRR Model)
 */
double calculate_european_option_price(const OptionParams *params) {
    // 1. Validations (Guard Clauses)
    if (params == NULL || params->steps <= 0 || params->time <= 0 || params->spot <= 0) {
        return -1.0;
    }

    int n = params->steps;
    double dt = params->time / n;
    double u = exp(params->volatility * sqrt(dt));
    double d = 1.0 / u;
    double a = exp(params->rate * dt);
    double p = (a - d) / (u - d);
    double q = 1.0 - p;

    // 2. Local Dynamic Memory Allocation (Thread Isolation)
    double *tree = (double *)malloc((n + 1) * sizeof(double));
    if (tree == NULL) return -1.0;

    // 3. Initialize Terminal Node Payoffs
    char type = tolower((unsigned char)params->type);
    for (int i = 0; i <= n; i++) {
        double stock_price = params->spot * pow(u, i) * pow(d, n - i);
        if (type == 'c') {
            tree[i] = stock_price > params->strike ? stock_price - params->strike : 0.0;
        } else {
            tree[i] = params->strike > stock_price ? params->strike - stock_price : 0.0;
        }
    }

    // 4. Backward Induction (O(n) Space Optimization)
    for (int j = n - 1; j >= 0; j--) {
        for (int i = 0; i <= j; i++) {
            tree[i] = (p * tree[i + 1] + q * tree[i]) / a;
        }
    }

    double final_price = tree[0];

    // 5. Explicitly Free Memory
    free(tree);

    return final_price;
}

File: src/option_pricing.py
import numpy as np
import math

def calculate_european_option_price(spot, strike, time, rate, volatility, steps, option_type='c'):
    """
    Calculates the European option price using the Cox-Ross-Rubinstein (CRR) model.
    Implemented with O(n) space complexity and vectorized operations using NumPy.
    
    Args:
        spot (float): Current stock price
        strike (float): Exercise price
        time (float): Time to maturity in years
        rate (float): Risk-free interest rate
        volatility (float): Annualized volatility
        steps (int): Number of binomial steps
        option_type (str): 'c' for Call, 'p' for Put
        
    Returns:
        float: Calculated option price
    """
    if steps <= 0 or time <= 0 or spot <= 0:
        return -1.0
        
    dt = time / steps
    u = math.exp(volatility * math.sqrt(dt))
    d = 1.0 / u
    a = math.exp(rate * dt)
    p = (a - d) / (u - d)
    q = 1.0 - p
    
    # Initialize terminal node stock prices
    # i goes from 0 to steps. Stock price = S * u^i * d^(steps - i)
    # Using vectorized NumPy powers for efficiency
    i = np.arange(steps + 1)
    stock_prices = spot * (u ** i) * (d ** (steps - i))
    
    # Initialize terminal payoffs
    if option_type.lower() == 'c':
        tree = np.maximum(stock_prices - strike, 0)
    else:
        tree = np.maximum(strike - stock_prices, 0)
        
    # Backward induction using in-place updates (O(n) space)
    # At each step j, we use the first j+1 elements of the array
    for j in range(steps - 1, -1, -1):
        # Vectorized step: tree[i] = (p * tree[i+1] + q * tree[i]) / a
        tree[:j+1] = (p * tree[1:j+2] + q * tree[:j+1]) / a
        
    return float(tree[0])
